src/autoschedulers/adams2019/cost_model_generator.cpp

```diff
@@ -15,38 +15,38 @@ using Halide::Derivative;
 // A model weight is either just an input, or an input and an output
 // (the updated weights and the ADAM state) depending on whether we're
 // doing inference or training.
-template<bool training>
```

Comment:
In this file, runtime dimensionality seems much cleaner (e.g. there's value in having all the types be the same down below)

Reply:
We could selectively revert this one. (I did the conversion by deleting all the old ctors temporarily and then fixing everything.)

Reply:
+1 to selectively reverting this one

Reply:
Before I revert this, can I ask what you mean by "having all the types be the same down below"? Looking at this again, I don't know which types you mean, and so I'm confused about the specific nature of your concern. (I'm trying to understand why you feel the old approach is cleaner, so I can factor different views into future work.)